mlir/lib/Conversion/MemRefToEmitC/MemRefToEmitC.cpp

```diff
@@ -77,6 +81,95 @@ struct ConvertAlloca final : public OpConversionPattern<memref::AllocaOp> {
   }
 };
 
+struct ConvertAlloc final : public OpConversionPattern<memref::AllocOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(memref::AllocOp allocOp, OpAdaptor operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    mlir::Location loc = allocOp.getLoc();
+    MemRefType memrefType = allocOp.getType();
+    if (!memrefType.hasStaticShape()) {
+      // TODO: Handle Dynamic shapes in the future. If the size
+      // of the allocation is the result of some function, we could
+      // potentially evaluate the function and use the result in the call to
+      // allocate.
+      return rewriter.notifyMatchFailure(
+          loc, "cannot transform alloc with dynamic shape");
+    }
+
+    mlir::Type sizeTType = mlir::emitc::SizeTType::get(rewriter.getContext());
+    Type elementType = memrefType.getElementType();
+    mlir::emitc::CallOpaqueOp sizeofElementOp =
+        rewriter.create<mlir::emitc::CallOpaqueOp>(
+            loc, sizeTType, rewriter.getStringAttr("sizeof"),
+            mlir::ValueRange{},
+            mlir::ArrayAttr::get(rewriter.getContext(),
+                                 {mlir::TypeAttr::get(elementType)}));
+    mlir::Value sizeofElement = sizeofElementOp.getResult(0);
+
+    unsigned int elementWidth = elementType.getIntOrFloatBitWidth();
+    IntegerAttr indexAttr = rewriter.getIndexAttr(elementWidth);
+
+    mlir::Value numElements;
+    numElements = rewriter.create<emitc::ConstantOp>(
+        loc, rewriter.getIndexType(), indexAttr);
+    mlir::Value totalSizeBytes = rewriter.create<emitc::MulOp>(
+        loc, sizeTType, sizeofElement, numElements);
+
+    emitc::CallOpaqueOp allocCall;
+    StringAttr allocFunctionName = allocOp.getAlignment()
+                                       ? rewriter.getStringAttr("aligned_alloc")
+                                       : rewriter.getStringAttr("malloc");
+    mlir::Value alignmentValue;
+    if (allocOp.getAlignment()) {
```

Comment:
nit: I wonder if you can do (I think it'd be more readable - because all the "has alignment" aspects would be in one place):

```
if (allocOp.getAlignment()) {
  allocFunctionName = ...
  alignmentValue = ...
} else {
  ...
}
```